### DISCUSSION TITLE
understanding about if (error) and try catch (error)

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#99] [其他問題] [answered]
以第四週 hw3 為例，我們會加上一些 if (error) 和 try catch 進行除錯。

* 問題一、為甚麼不將欲執行的程式碼全部包在 try 裡面，不符合的情況就會跳去 catch 中 ? 是因為除錯的時候 catch 中的狀況太多，會不曉得是哪一種狀況出錯，所以不這麼做嗎?
``` js
try{
 // 欲執行的程式碼
}catch(e){
 // 當錯誤發生時，欲執行的程式碼
}
```
``` js
const request = require('request')

const process = require('process')

const name = process.argv[2]

request(
  `https://restcountries.eu/rest/v2/name/${name}`,

  (error, response, body) => {
    const json = JSON.parse(body)
    // eslint-disable-next-line
    const length = Object.keys(json).length
    try {
      for (let i = 0; i < length; i++) {
        console.log('============')
        console.log(`國家 : ${json[i].name}`)
        console.log(`首都 : ${json[i].capital}`)
        console.log(`貨幣 : ${json[i].currencies[0].code}`)
        console.log(`國碼 : ${json[i].callingCodes}`)
      }
    } catch (error) {
      console.log('找不到國家資訊')
    }
  }

)
```
* 問題二、 if (error) 的 error 是 status code 的 error (4開頭的)、而 catch (error) 這邊的 error 是針對不符合 try 情況所導致的error對嗎?  名稱相同但兩邊的參數所代表的是不同的意思，我的理解對嗎? 

感謝大家的幫忙及回答~

**Comment by aszx87410** [comment, marked as answer]
> 問題一、為甚麼不將欲執行的程式碼全部包在 try 裡面，不符合的情況就會跳去 catch 中 ? 是因為除錯的時候 catch 中的狀況太多，會不曉得是哪一種狀況出錯，所以不這麼做嗎?

有些地方你很確定不會出錯，那就沒有放的必要，如果全部都包在裡面的話通常就代表寫的人也不知道哪邊有可能出錯，是不好的習慣

try catch 基本上是用在：我預期這邊會有錯誤，而我知道要怎麼處理這錯誤

> 問題二、 if (error) 的 error 是 status code 的 error (4開頭的)、而 catch (error) 這邊的 error 是針對不符合 try 情況所導致的error對嗎? 名稱相同但兩邊的參數所代表的是不同的意思，我的理解對嗎?

前者錯了，後者是對的

if (error) 的 error 不一定是 status code 的 error，而是 request 這個 library 執行動作時可能發生的錯誤，至於詳細的定義是什麼，要看 request 內部的實作是什麼

**Comment by allenliaooooooo** [comment]
> * 問題二、 if (error) 的 error 是 status code 的 error (4開頭的)、而 catch (error) 這邊的 error 是針對不符合 try 情況所導致的error對嗎?  名稱相同但兩邊的參數所代表的是不同的意思，我的理解對嗎?

`if(error)` 的 `error`，是 `request()` 這個方法自己有一個錯誤處理，當發生錯誤的時候 `error` 就會有東西，但實際是什麼我也不清楚，因為沒遇過，看了文件好像也沒有說明，有可能是簡單的 string 寫著 error message，也有可能是一個 Error 物件，取決於 `request()` 這個方法是怎麼寫的
`catch(error)` 是如果在 try 裡面發生錯誤，可能是程式沒有寫好的錯誤（我們常在 console 裡面看到的錯誤），你也可以自己寫錯誤處理，自己丟一個 Error 物件，然後 `catch` 就會幫你接住這個錯誤，然後執行你寫在 `catch` 裡面的事情
以上是我自己淺薄的理解，如果有錯再麻煩其他人糾正一下🙏

**Comment by s103071049** [comment]
> if (error) 的 error 不一定是 status code 的 error，而是 request 這個 library 執行動作時可能發生的錯誤，至於詳細的定義是什麼，要看 request 內部的實作是什麼


所以，綜合老師以及 rockyooooooo 大大的說明，我以第四週 hw1 來為例，解釋我對 request  這個 library 執行動作時可能發生錯誤的理解。以下的理解正確嗎?

我送一個 request 過去，server 傳一個 response 回來。如果這個執行過程有 error ，就會跑到 error 這個參數當中，而下面程式碼做的只是將如果出現 error 的 error 印出來；但是如果發出 request 到回傳 response 這個過程的執行是正常，我想對回傳後的內容做檢查，像是他的回傳格式是不是我要的，如果是我要的，我再做下列運算，如果不是則將他的錯誤印出來，所以就出現了 try catch 的檢查，而 catch 裡面的錯誤並不算是 request 這個 library 執行動作時可能發生的錯誤，我的理解對嗎? 
 

``` js
const request = require('request')

request(
  'https://lidemy-book-store.herokuapp.com/books?_limit=10',
  (error, response, body) => {
    
    if (error) {
      return console.log('error', error)
    }

    
    let books
    try {
      books = JSON.parse(body)
    } catch (error) {
      return console.log('error', error)
    }
    for (let i = 0; i < books.length; i++) {
      console.log(`${books[i].id} ${books[i].name}`)
    }
  })

```

**Reply by allenliaooooooo** [reply]
> 我送一個 request 過去，server 傳一個 response 回來。如果這個執行過程有 error ，就會跑到 error 這個參數當中，而下面程式碼做的只是將如果出現 error 的 error 印出來；

不太確定你說 *就會跑到 error 這個參數當中* 是什麼意思
我自己的說法會是如果 `request` 的過程有錯誤發生，就會把錯誤傳到 `error` 這個參數
然後 `response` 跟 `body` 可能還是會有東西（看 `request()` 內部怎麼寫）

> 但是如果發出 request 到回傳 response 這個過程的執行是正常，我想對回傳後的內容做檢查，像是他的回傳格式是不是我要的，如果是我要的，我再做下列運算，如果不是則將他的錯誤印出來，所以就出現了 try catch 的檢查，而 catch 裡面的錯誤並不算是 request 這個 library 執行動作時可能發生的錯誤，我的理解對嗎?

這邊應該沒錯，有錯再請其他人幫忙說明

-----
可以參考看看[第四週挑戰題老師給的範例解答](https://github.com/Lidemy/mentor-program-5th-rockyooooooo/blob/master/examples/week4/twitch2.js)
可以把我們在用的 `request()` 想像成老師在裡面寫的 `getAllStream()`
發生錯誤會把 `error` 丟給 callback，然後 callback 就是我們在呼叫 `getAllStream()` 時候傳進去的匿名函式
我想基本上 `request()` 大概就是這個形式

**Reply by s103071049** [reply]
非常感謝 @rockyooooooo 大大的幫忙~
